### PR TITLE
Convert Stl to a class created via constructor, separate out error types.

### PR DESCRIPTION
**Pull request by DJankauskas**
Allow creating `Stl` instances more idiomatically. Since error objects have no reliance on stl context, further separates error types out of `Stl`.

### CODE REVIEW

File: packages/stainless/src/stl.ts [hunk 1]
```diff
@@ -267,288 +267,200 @@ export type OpenAPIEndpoint = Endpoint<
...
-      const openapiEndpoint = openapi?.endpoint ?? "get /api/openapi";
-      const topLevelActions = topLevel?.actions || {};
-      const apiDescription = {
-        openapi: {
-          endpoint: openapiEndpoint,
-          get spec() {
-            // TODO memoize
-            return stl.openapiSpec(apiDescription);
-          },
+  }): APIDescription<TopLevel & (typeof openapi extends { endpoint: false } ? {} : { actions: { getOpenapi: OpenAPIEndpoint } }), Resources> {
```
Comment (DJankauskas) [inline packages/stainless/src/stl.ts:434]: This type is ugly, is there a nicer way to do this?

**Comment by DJankauskas** [comment]
Consideration for future work: are the `make*` functions for creating plugins idiomatic, and should they be changed somehow? They return functions, so the naming does make more sense.

**Review by jedwards1211 — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -31,7 +31,7 @@ const plugins = {
   prisma: makePrismaPlugin(),
 };
 
-export const stl = makeStl<{}, typeof plugins>({
+export const stl = new Stl<{}, typeof plugins>({
   plugins,
 });
```
Comment (rattrayalex) [inline README.md:36]: I wonder if there's a way to not make the user pass a generic type here? having to pass `plugins` and then also `typeof plugins` feels a little off…

File: demo/libs/currentUserPlugin.ts [hunk 3]
```diff
@@ -19,7 +20,7 @@ export const makeCurrentUserPlugin =
       context.requireCurrentUser = (): User => {
         const { currentUser } = context;
         if (!currentUser) {
-          throw new stl.UnauthorizedError({ naughty: "you aint logged in!" });
+          throw new UnauthorizedError({ naughty: "you aint logged in!" });
```
Comment (rattrayalex) [inline demo/libs/currentUserPlugin.ts:23]: Why these changes?

**Comment by rattrayalex** [comment]
> Consideration for future work: are the make* functions for creating plugins idiomatic, and should they be changed somehow? They return functions, so the naming does make more sense.

Honestly that's a good question, and I don't know the answer – what do you find when you look at other plugin systems?

File: README.md (hunk 2 above)
Comment (DJankauskas) [reply README.md:36]: It's possible to do that, but it reduces the flexibility of the framework for advanced use-cases. I have a local branch that implements this, but Andy and I aren't sure if it's the right way forward, your input would be very useful here. The solution is that instead of the user passing in a `UserContext` type, if they wish to add context, they extend an `StlUserContext` interface.
